monitor.js

```diff
@@ -0,0 +1,94 @@
+const pull = require('pull-stream')
+
+function isForeignLeafFeedBranch(branch) {
+  if (branch.length !== 4) return false
+  if (branch[0].keys) return false // not interested in my branches
+  if (branch[0].purpose !== 'root') return false
+  if (branch[1].purpose !== 'v1') return false
+  if (branch[2].purpose.length !== 1) return false
+  return true
+}
+
+/**
+ * There may have been messages I had in my database but was unable to decrypt
+ * because I didn't know my root metafeed keys. Now that I know them, I should
+ * reindex my database and decrypt all messages.
+ *
+ * Then, we should restart monitoring because we have the root and now we can
+ * finally create DM keys for foreign leafs.
+ */
+function reindexAndRestart(ssb, encryptionFormat, drain) {
+  if (drain) drain.abort()
+  ssb.db.reindexEncrypted((err) => {
+    if (err) console.warn(err)
+
+    monitorForest(ssb, encryptionFormat)
+  })
+}
+
+/**
+ * Listen to a stream of all branches in the metafeed "forest" (collection of
+ * trees) and send some information such as keys to the box2 encryptionFormat.
+ *
+ * Box2 encryptionFormat uses keys from the metafeed tree to encrypt messages,
+ * this is why we need this information
+ */
+function monitorForest(ssb, encryptionFormat) {
+  encryptionFormat.getRootSigningKey((err, rootKeys) => {
+    if (err) console.warn(err)
```

Comment:
Might want to add a comment here :-) I thought it would error, but now that I read it again I see that this is the "I don't have the root yet" case.

Reply:
wait, that function will never err

Reply:
Fixed